File: sdk-python/cak_sdk/transport.py
"""线协议（与 sdk/transport.ts 对齐）：未知 cak 版本 → -32600；未知 method → -32601；坏 JSON → -32700。"""
from __future__ import annotations
import json
from typing import Any, Callable, Optional, Union

CAK_ENVELOPE_VERSION = "1"

class RPC:
    PARSE = -32700
    INVALID_REQUEST = -32600
    METHOD_NOT_FOUND = -32601
    INVALID_PARAMS = -32602
    INTERNAL = -32603
    CANCELLED = -32800

METHODS = ("plugin.hello", "plugin.health", "plugin.shutdown", "capability.execute", "model.generate", "context.contribute", "interceptor.intercept", "cancel", "event.publish")

def encode(envelope: dict) -> str:
    return json.dumps(envelope, ensure_ascii=False, separators=(",", ":")) + "\n"

def decode(line: str) -> Union[dict, "DecodeError"]:
    """合法信封 → dict；否则 DecodeError（带 id 以便回错误）"""
    try:
        o = json.loads(line)
    except Exception:
        return DecodeError(RPC.PARSE, "parse error", None)
    if not isinstance(o, dict):
        return DecodeError(RPC.INVALID_REQUEST, "not an object", None)
    rid = o.get("id") if isinstance(o.get("id"), (int, str)) else None
    if o.get("cak") != CAK_ENVELOPE_VERSION:
        return DecodeError(RPC.INVALID_REQUEST, f"unsupported envelope version {o.get('cak')!s} (expected {CAK_ENVELOPE_VERSION})", rid)
    if o.get("jsonrpc") != "2.0":
        return DecodeError(RPC.INVALID_REQUEST, "jsonrpc must be 2.0", rid)
    return o

class DecodeError:
    __slots__ = ("code", "message", "id")
    def __init__(self, code: int, message: str, rid: Optional[Union[int, str]]):
        self.code, self.message, self.id = code, message, rid

def request(rid: Union[int, str], method: str, params: dict) -> dict:
    return {"cak": CAK_ENVELOPE_VERSION, "jsonrpc": "2.0", "id": rid, "method": method, "params": params}

def response(rid: Optional[Union[int, str]], result: Any) -> dict:
    return {"cak": CAK_ENVELOPE_VERSION, "jsonrpc": "2.0", "id": rid, "result": result}

def failure(rid: Optional[Union[int, str]], code: int, message: str, data: Optional[dict] = None) -> dict:
    err: dict = {"code": code, "message": message}
    if data:
        err["data"] = data
    return {"cak": CAK_ENVELOPE_VERSION, "jsonrpc": "2.0", "id": rid, "error": err}

class LineSplitter:
    """stdin 可能一次给半行 / 多行；按 \\n 切、去空行"""
    def __init__(self) -> None:
        self._buf = ""
    def push(self, chunk: str, on_line: Callable[[str], None]) -> None:
        self._buf += chunk
        while True:
            i = self._buf.find("\n")
            if i < 0:
                break
            line = self._buf[:i].strip()
            self._buf = self._buf[i + 1:]
            if line:
                on_line(line)

File: sdk/transport.ts
/**
 * @cak/sdk · 传输协议（09 §C，M3）：JSON-RPC 2.0 over NDJSON（stdio），信封 { cak:"1", jsonrpc:"2.0", ... }。
 * 未知 cak 版本 → -32600；未知 method → -32601（显式，不静默）。
 * 本文件是插件端与内核端共用的纯协议工具（无内核依赖）。
 */
import type { Json, JsonObject } from './types.js';

export const CAK_ENVELOPE_VERSION = '1';
export interface Envelope { cak: '1'; jsonrpc: '2.0'; id?: number | string | null; method?: string; params?: JsonObject; result?: Json; error?: { code: number; message: string; data?: JsonObject } }
export const RPC = { PARSE: -32700, INVALID_REQUEST: -32600, METHOD_NOT_FOUND: -32601, INVALID_PARAMS: -32602, INTERNAL: -32603, CANCELLED: -32800 } as const;

export const METHODS = ['plugin.hello', 'plugin.health', 'plugin.shutdown', 'capability.execute', 'model.generate', 'context.contribute', 'interceptor.intercept', 'cancel', 'event.publish',
  // N-48（只增）：控制器跑子进程——内核→插件 controller.decide{decideId, view, config}；决策期间插件→内核反向请求 ctx.*{decideId,…}
  'controller.decide', 'ctx.invoke', 'ctx.compose', 'ctx.preview', 'ctx.attenuate', 'ctx.spawn'] as const;
export type Method = typeof METHODS[number];

export const encode = (e: Envelope) => JSON.stringify(e) + '\n';
export function decode(line: string): Envelope | { error: { code: number; message: string }; id?: number | string | null } {
  let o: any; try { o = JSON.parse(line); } catch { return { error: { code: RPC.PARSE, message: 'parse error' } }; }
  if (!o || typeof o !== 'object') return { error: { code: RPC.INVALID_REQUEST, message: 'not an object' } };
  const id = typeof o.id === 'number' || typeof o.id === 'string' ? o.id : null;
  if (o.cak !== CAK_ENVELOPE_VERSION) return { error: { code: RPC.INVALID_REQUEST, message: `unsupported envelope version ${String(o.cak)} (expected ${CAK_ENVELOPE_VERSION})` }, id };
  if (o.jsonrpc !== '2.0') return { error: { code: RPC.INVALID_REQUEST, message: 'jsonrpc must be 2.0' }, id };
  return o as Envelope;
}
export const request = (id: number | string, method: Method, params: JsonObject): Envelope => ({ cak: '1', jsonrpc: '2.0', id, method, params });
export const response = (id: number | string | null, result: Json): Envelope => ({ cak: '1', jsonrpc: '2.0', id, result });
export const failure = (id: number | string | null, code: number, message: string, data?: JsonObject): Envelope => ({ cak: '1', jsonrpc: '2.0', id, error: { code, message, ...(data ? { data } : {}) } });

/** 按行切分的流式解析器（stdin 可能一次给半行 / 多行） */
export class LineSplitter {
  private buf = '';
  push(chunk: string, onLine: (line: string) => void) { this.buf += chunk; let i; while ((i = this.buf.indexOf('\n')) >= 0) { const line = this.buf.slice(0, i).trim(); this.buf = this.buf.slice(i + 1); if (line) onLine(line); } }
}
